gadget/install/partition.go

```diff
@@ -264,6 +265,28 @@ func removeCreatedPartitions(gadgetRoot string, lv *gadget.LaidOutVolume, dl *ga
 	return nil
 }
 
+func partitionsWithRolesAndContent(lv *gadget.LaidOutVolume, dl *gadget.OnDiskVolume, roles []string) []gadget.OnDiskStructure {
+	roleForOffset := map[quantity.Offset]*gadget.LaidOutStructure{}
+	for idx, gs := range lv.LaidOutStructure {
+		if gs.Role != "" {
+			roleForOffset[gs.StartOffset] = &lv.LaidOutStructure[idx]
+		}
+	}
+
+	var parts []gadget.OnDiskStructure
+	for _, part := range dl.Structure {
+		gs := roleForOffset[part.StartOffset]
+		if gs == nil || gs.Role == "" || !strutil.ListContains(roles, gs.Role) {
+			continue
+		}
+		// now that we have a match, override the laid out structure
+		// such that the fields reflect what was declared in the gadget
```

Comment:
because of this behavior, if I understand correctly it means if something was resized we get the size as per gadget? this function itself probably merits a doc comment clarifying this. OTOH I'm nor sure if offset and size are even used by the code in FactoryReset?

Reply:
Offset is not but the size is, fortunately not in a way that would create a smaller partition, it's only used to apply some default sector size optimization for smaller than 32MB partitions. This is to have a similar behavior as createMissingPartitions() which unfortunately mixes gadget.OnDiskStructure with laid out structure which then, which later is consumed by mkfs/write content helpers. I'll add a comment though.

Reply:
Actually I forgot that OnDiskStructure already has a Size field. In the corresponding code in createMissingPartitions() we update that field, while the value from gadget as set in LaidOutStructure is never modified.